README.md
This is an extremely complicated and almost completely useless browser sniffing library written in JavaScript. Useless because you shouldn't use browser sniffing. So stop right now and go read something about feature detecting instead. I'm serious. Go away. You'll thank me later.

![Build Status](https://github.com/WhichBrowser/Parser-JavaScript/actions/workflows/main.yml/badge.svg?event=push)
[![npm](https://img.shields.io/npm/l/which-browser.svg)](https://www.npmjs.com/package/which-browser)
[![npm](https://img.shields.io/npm/v/which-browser.svg)](https://www.npmjs.com/package/which-browser)
[![Twitter Follow](https://img.shields.io/twitter/follow/simariot.svg?style=social)](https://twitter.com/simariot)

---

If you are looking for the PHP version of WhichBrowser, please go to the [WhichBrowser/Parser-PHP](https://github.com/WhichBrowser/Parser-PHP) project.

If you are looking for the server written in PHP that provides a Javascript API for the browser, please go to the [WhichBrowser/Server](https://github.com/WhichBrowser/Server) project.

---

**But why _almost completely useless_ and not completely useless?**  
Well, there is always an exception to the rule. There are valid reasons to do browser sniffing: to improve the user experience or to gather intelligence about which browsers are used on your website. The original author website is html5test.com and he wanted to know which score belongs to which browser. And to do that you need a browser sniffing library.

**Why is it extremely complicated?**  
Because everybody lies. Seriously, there is not a single browser that is completely truthful. Almost all browsers say they are Netscape 5 and almost all WebKit browsers say they are based on Gecko. Even Internet Explorer 11 now no longer claims to be IE at all, but instead an unnamed browser that is like Gecko. And it gets worse. That is why it is complicated.

**What kind of information does it give?**
You get a nice object which has information about the browser, rendering engine, os and device. It gives you names and versions and even device manufacturer and model. And WhichBrowser is pretty tenacious. It gives you info that others don't. For example:

    JUC (Linux; U; 2.3.6; zh-cn; GT-I8150; 480*800) UCWEB8.7.4.225/145/800
    UC Browser 8.7 on a Samsung Galaxy W running Android 2.3.6

Android is never mentioned

    Mozilla/5.0 (Series40; Nokia501/10.0.2; Profile/MIDP-2.1 Configuration/CLDC-1.1) Gecko/20100401 S40OviBrowser/3.0.0.0.73
    Nokia Xpress 3.0.0 on a Nokia Asha 501 running Nokia Asha Platform

Despite the useragent header claiming to be a Series40 device, we know it's actually running the Asha Platform and we also know that OviBrowser has been renamed to Nokia Xpress.

    Opera/9.80 (X11; Linux zvav; U; zh) Presto/2.8.119 Version/11.10
    Opera Mini on a Nokia 5230 running Series60 5.0

The useragent header looks like Opera 11.10 on Linux, but we know it's Opera Mini. We can even figure out the real operating system and device model from other headers.

## Requirements

WhichBrowser requires Node 6 or higher to run and Node 16 or higher to build and test.

## Dependencies

WhichBrowser has no dependencies, is just plain ES6.

## How to install it

You can install WhichBrowser by NPM and Yarn

    npm install which-browser

or

yarn add which-browser

Do not stick to a specific WhichBrowser version, otherwise you will miss all the sniffing engine updates

## How to use it (short version)

```js
const WhichBrowser = require("which-browser");
const http = require("http");
const port = 3000;

const server = http.createServer((request, response) => {
  const result = new WhichBrowser(request.headers);
  response.end(result.toString());
});

server.listen(port, (err) => {
  if (err) {
    return console.log("Something is broken ¯\\_(ツ)_/¯", err);
  }
  console.log(`Server is listening on port ${port} ✌(-‿-)✌`);
});
```

## How to use it (long version)

The first step require WhichBrowser:

```js
const WhichBrowser = require("which-browser");
```

The second step is to create a new instance of the `WhichBrowser` class. This object will contain all the information the library could find about the browser. The object has a required parameter, either the headers send by the browser, or a useragent string. Using the request headers is preferable, because it will allow a better detection, but if you have just the useragent string, this will also work.

For example:

```js
const result = new WhichBrowser(request.headers);
```

or:

```js
const result = new WhichBrowser(request.headers['user-agent']));
```

The variable `result` now contains an object which you can query for information. There are various ways to access the information.

First of all, you can call to `toString()` function to get a human readable identification:

```js
`You are using ${result.toString()}`;
// You are using Chrome 27 on OS X Mountain Lion 10.8
```

Another possiblity is to query the object:

```js
result.isType("desktop");
// true

result.isType("mobile", "tablet", "media", "gaming:portable");
// false

result.isBrowser("Maxthon", "<", "4.0.5");
// false

result.isOs("iOS", ">=", "8");
// false

result.isOs("OS X");
// true

result.isEngine("Blink");
// true
```

You can also access these properties directly:

```js
result.browser.toString();
// Chrome 27

result.engine.toString();
// Blink

result.os.toString();
// OS X Mountain Lion 10.8
```

Or access parts of these properties directly:

```js
result.browser.name;
// Chrome

`${result.browser.name} ${result.browser.version.toString()}`;
// Chrome 27

result.browser.version.value;
// 27.0.1453.110

result.engine.name;
// Blink
```

Finally you can also query versions directly:

```js
result.browser.version.is(">", 26);
// true

result.os.version.is("<", "10.7.4");
// false
```

## Options

It is possible to set additional options by passing an object as the second parameter when creating the `WhichBrowser` object.

### Disabling detection of bots

In some cases you may want to disable the detection of bots. This allows the bot the deliberately fool WhichBrowser, so you can pick up the identity of useragent what the bot tries to mimic. This is especially handy when you want to use WhichBrowser to switch between different variants of your website and want to make sure crawlers see the right variant of the website. For example, a bot that mimics a mobile device will see the mobile variant of you site.

```js
result = new WhichBrowser(request.headers, { detectBots: false });
```

## Enable result caching

WhichBrowser can cache results between requests. Using a cache is especially useful if you use WhichBrowser on every page of your website and a user visits multiple pages. During the first visit the headers will be parsed and the result will be cached. Upon further visits, the cached results will be used, which is much faster than having to parse the headers again and again.

In order to enable caching you need to specify in the options object which type of cache you want use.

For example, if you want to enable the cache you need to construct the `WhichBrowser` object in this way:

```js
const result = new WhichBrowser(request.header, {
  cache: WhichBrowser.SIMPLE_CACHE,
});
```

If a result is retrieved from the cache it has the `cached` property set to `true`

```js
if (result.cached) {
  // from cache
} else {
  // just parsed for the first time
}
```

You can also specify after how many seconds a cached result should be discarded. The default value is 900 seconds or 15 minutes. If you think WhichBrowser uses too much memory for caching, you should lower this value. You can do this by setting the `cacheExpires` property in the options object.

For example, if you want that your cached results lasts for 300 seconds or 5 minutes do:

```js
const result = new WhichBrowser(request.header, {
  cache: WhichBrowser.SIMPLE_CACHE,
  cacheExpires: 300,
});
```

A value for `cacheExpires` less or equal to `0` disable the expiry for that result and it will last until you restart node or you parse the same set of headers with a `cacheExpires` greater than `0`.

Cache validity is checked at a rate of `cacheExpires / 5` so, with a `cacheExpires` of `500`, you can rest assured that your result has been reaped from the cache after `500 + 500 / 5 + 1` seconds.

If you want to speed up the process of validity check you can set the `cacheCheckInterval` property in the options object. This property can't be smaller than `1`.

```js
const result = new WhichBrowser(request.header, {
  cache: WhichBrowser.SIMPLE_CACHE,
  cacheExpires: 300,
  cacheCheckInterval: 1,
});
```

In this way the cache lasts for `300` seconds but is checked every `1` second.

> Be aware that changing `cacheExpiries` or `cacheCheckInterval` impact the cache validity check rate for **ALL** records in the cache. For example setting `cacheExpiries` to `0` will prevent **ALL** results to expire because it will disable the cache validity check (for the sake of truth it will be done every `57085` years, `5` months, `10` days, `7` hours, `35` minutes and `48` seconds).

## API reference

### The WhichBrowser object

After a new `WhichBrowser` object is created, it contains a number of properties and functions. All of these properties are guaranteed to be present.

**Properties:**

- `browser`  
  an object that contains information about the browser itself
- `engine`  
  an object that contains information about the rendering engine
- `os`  
  an object that contains information about the operating system
- `device`  
  an object that contains information about the device

**Functions:**

`getType()`  

CHANGELOG.md
#### 0.7.1 (2023-11-01)

- **Database:** Added Chrome 101 to 118.

#### 0.7.0 (2022-04-21)

- Sync WhichBrowser/Parser-JavaScript to [WhichBrowser/Parser-PHP@1044880](https://github.com/WhichBrowser/Parser-PHP/commit/1044880bc792dbce5948fbff22ae731c43c280d9)

- Supports [Chrome UA reduction](https://developer.chrome.com/docs/privacy-sandbox/user-agent/)

- **BREAKING:** `Plasystation` has been reworded to `PlayStation` across the whole codebase. 

- **Database:** 
  * Adds support for `Huawei Browser`
  * Adds support for `Xbox Series X`
  * Adds support for `PlayStation 5`
  * Adds support for `HarmonyOS`



#### 0.6.1 (2022-04-07)

- **Database:** Added Chrome 99 and 100 ([a770c30](https://github.com/WhichBrowser/Parser-JavaScript/commit/a770c30d29045893797b79f6c623b9ab8ebb8cc0))
- **Fixes:** Chrome 100+ was being detected as a browser in Camouflage mode. Fixes [#61](https://github.com/WhichBrowser/Parser-JavaScript/issues/61)

#### 0.6.0 (2022-02-16)

- **Database:** Synced the JavaScript version of WhichBrowser with the latest changes happened in WhichBrowser-PHP until [WhichBrowser-PHP@d082883](https://github.com/WhichBrowser/Parser-PHP/commit/d0828833c1e2e05654b72e78d6944732adeb606b) [[diff](https://github.com/WhichBrowser/Parser-PHP/compare/0ef1a43950f14c65a760fe8123b4401e2b315bc3...d082883)].
- **Operations:** Migrate from TravisCI to GitHub Actuibs

#### 0.5.1 (2019-11-30)

- **Database:** Added Chrome 77 and 78.

#### 0.5.0 (2019-08-25)

- **Database:** Added support for Chromium based Edge. Updated Android and iOS models. Added Chrome 75 and 76. Added Samsung Galaxy Fold and Galaxy A models.
- **Fixes:** False positive for Obigo browser.

#### 0.4.2 (2019-06-01)

Now _TypeScript Types_ are available thanks to @[AnandChowdhary](https://github.com/AnandChowdhary) and his PR [#36](https://github.com/WhichBrowser/Parser-JavaScript/pull/36)

- **Database:** Added Chrome 73 and 74 ([d2ea800](https://github.com/WhichBrowser/Parser-JavaScript/commit/d2ea8009277426ea874559e8e426f29510816ae0))

#### 0.4.0 (2019-02-07)

- **Fix:** [Issue 96 of WhichBrowser/Parser-PHP](https://github.com/WhichBrowser/Parser-PHP/issues/96), `profiles.js` and `models-*.js` files had some occourrences of spaces replaced by Unicode Character 'EN SPACE' (U+2002). The scripts that download these files are now stripping away that character and replacing it with a normal space. Thanks @[koconder](https://github.com/koconder) for [#29](https://github.com/WhichBrowser/Parser-JavaScript/pull/29) and @[Taiwaninja](https://github.com/Taiwaninja) for reporting it.
- **Tools:** WhichBrowser/Parser-Javascript is now tested on all the node versions between 6 and 11 included. Thanks again to @[koconder](https://github.com/koconder) for [#30](https://github.com/WhichBrowser/Parser-JavaScript/pull/30).
- **Database:** Added Chrome 70, 71 and 72 ([34735c2](https://github.com/WhichBrowser/Parser-JavaScript/commit/34735c27191e6b4c70b850aeea6f90bc51c65299))

#### 0.3.5 (2018-10-01)

- **Database:** Added new iPhones XS, XS Max, XR ([911ae7a](https://github.com/WhichBrowser/Parser-JavaScript/commit/911ae7a7847acc2be01a809db95bbfaeb930ae73))

#### 0.3.4 (2018-09-13)

- **Database:** Added Chrome 69 ([5b8a4fa](https://github.com/WhichBrowser/Parser-JavaScript/commit/5b8a4fa5e81740336cf29c88982af152f70e81c9))

#### 0.3.3 (2018-08-22)

- **Fixes:** Now is possible to _really_ disable the bot detection with the option `detectBots: false` ([5e02e9a](https://github.com/WhichBrowser/Parser-JavaScript/commit/5e02e9a01fdee83b7bb0b6e91372b1870b157d52))

#### 0.3.2 (2018-07-27)

- **Database:** Added Chrome 68 ([0d04c8d](https://github.com/WhichBrowser/Parser-JavaScript/commit/0d04c8def9971c7a6bf16e1b1398bb475e69a310) & [cd5ec81](https://github.com/WhichBrowser/Parser-JavaScript/commit/cd5ec8183a6145dd901e3433c78846e1f9af5976))
- **Database:** Fixed name of Oculus Browser and added Oculus Go detection, thanks @frankolivier (#23) ([b38377c](https://github.com/WhichBrowser/Parser-JavaScript/commit/b38377c8d1856a316c6f74fea15ee3228aa050d6))

#### 0.3.1 (2018-07-13)

- **Database:** Updated Models, Indices and Profiles ([fa4dd85](https://github.com/WhichBrowser/Parser-JavaScript/commit/fa4dd8514991578da4102027269071db860194f3))

#### 0.3.0 (2018-06-19)

- **Database:** Added Chrome 66 & 67 ([a9bce87](https://github.com/WhichBrowser/Parser-JavaScript/commit/a9bce872a9a17491539aca760d2627f1af2a6beb))

#### 0.2.9 (2018-03-16)

- **Database:** Added Chrome 65 ([c1c9148](https://github.com/WhichBrowser/Parser-JavaScript/commit/c1c9148397b179d0655ec0f5b181d8e65511df38))

#### 0.2.8 (2018-03-06)


data/models-wp.js
/* This file is automatically generated, do not edit manually! */

/* eslint-disable */

const DeviceType = require('../src/constants').deviceType;

exports.WP_MODELS = {
  Allegro: ['Acer', 'Allegro'],
  M310: ['Acer', 'Allegro'],
  S58: ['Acer', 'Jade Primo'],
  'M220!': ['Acer', 'Liquid'],
  FierceXL: ['Alcatel', 'One Touch Fierce XL'],
  'ONE TOUCH 5040X': ['Alcatel', 'One Touch View'],
  'IDOL 4 PRO': ['Alcatel', 'Idol 4 Pro'],
  'IDOL 4S with Windows 10': ['Alcatel', 'Idol 4S'],
  'Impera I': ['Allview', 'Impera i'],
  'Impera M': ['Allview', 'Impera M'],
  'Impera S': ['Allview', 'Impera S'],
  W1i: ['Allview', 'W1i'],
  '40 Cesium': ['Archos', '40 Cesium'],
  AC50CE: ['Archos', '50 Cesium'],
  '- BUSH Windows Phone': ['Argos', 'BUSH Windows Phone'],
  Galaxy6: ['Asus', 'Galaxy 6'],
  E600: ['Asus', 'E600'],
  'WIN HD LTE': ['BLU', 'Win HD'],
  'WIN HD W510l': ['BLU', 'Win HD'],
  'WIN HD W510u': ['BLU', 'Win HD'],
  'WIN JR LTE': ['BLU', 'Win JR'],
  'WIN JR W410a': ['BLU', 'Win JR'],
  'WIN JR W410i': ['BLU', 'Win JR'],
  'WIN JR W410l': ['BLU', 'Win JR'],
  'WIN JR W410u': ['BLU', 'Win JR'],
  'BP30.BM180': ['Bluebird', 'Pidion'],
  'WIN 400': ['Celkon', 'Win 400'],
  'Alpha Neon': ['Cherry Mobile', 'Alpha Neon'],
  'ALPHA PRIME 5': ['Cherry Mobile', 'Alpha Prime 5'],
  'MOBILE Alpha Luxe': ['Cherry Mobile', 'Alpha Luxe'],
  'CHERRY-MOBILE Alpha Luxe': ['Cherry Mobile', 'Alpha Luxe'],
  'MOBILE Alpha Style': ['Cherry Mobile', 'Alpha Style'],
  'MOBILE Alpha View': ['Cherry Mobile', 'Alpha View'],
  'PKT-407': ['Condor', 'Griffe W1'],
  T698: ['Cube', 'WP10', DeviceType.TABLET],
  'Venue Pro': ['Dell', 'Venue Pro'],
  'Ixion W 5': ['DEXP', 'Ixion W 5'],
  IQ400W: ['Fly', 'IQ400W Era Windows'],
  IQ500W: ['Fly', 'IQ500W Evo Windows'],
  FTJ152E: ['Freetel', 'KATANA 01'],
  FTJ152F: ['Freetel', 'KATANA 02'],
  IS12T: ['Fujitsu Toshiba', 'IS12T'],
  W10: ['GoMobile', 'W10'],
  WINJOY: ['Highscreen', 'WinJoy'],
  WinWin: ['Highscreen', 'WinWin'],
  'E260T+': ['Hisense', 'NANA'],
  'SG 7': ['HIWIKI', 'SG 7'],
  DolphinCT50: ['Honeywell', 'Dolphin CT50'],
  '75eL0N': ['Honeywell', 'Dolphin 75e'],
  HELSINKI: ['Honeywell', '"Helsinki"'],
  HD2: ['HTC', 'HD2'],
  ';HD2': ['HTC', 'HD2'],
  LEO: ['HTC', 'HD2'],
  'HD2(Leo': ['HTC', 'HD2'],
  'HD2 LEO': ['HTC', 'HD2'],
  'HTC HD2': ['HTC', 'HD2'],
  '(HD7|.*T92(92|95|96))!': ['HTC', 'HD7'],
  'Schubert!': ['HTC', 'HD7'],
  Mondrian: ['HTC', 'HD7'],
  Gold: ['HTC', 'HD7'],
  'Touch-IT HD7': ['HTC', 'HD7'],
  '((7 )?Mozart|.*T86(97|98))!': ['HTC', '7 Mozart'],
  PD67100: ['HTC', '7 Mozart'],
  ';PD67100': ['HTC', '7 Mozart'],
  'HTC Mozart': ['HTC', '7 Mozart'],
  '7 HTC MOZART': ['HTC', '7 Mozart'],
  PC93100: ['HTC', '7 Pro'],
  MWP6885: ['HTC', '7 Pro'],
  '(7 Pro|.*T75(75|76))!': ['HTC', '7 Pro'],
  'USCCHTC-PC93100': ['HTC', '7 Pro'],
  Surround: ['HTC', '7 Surround'],
  '7 Surround': ['HTC', '7 Surround'],
  '.*T8788!': ['HTC', '7 Surround'],
  '((7 )?Trophy|.*T8686)!': ['HTC', '7 Trophy'],
  mwp6985: ['HTC', '7 Trophy'],
  'htc mpw6958': ['HTC', '7 Trophy'],
  htcmpw6958: ['HTC', '7 Trophy'],
  PC40100: ['HTC', '7 Trophy'],
  PC40200: ['HTC', '7 Trophy'],
  'Touch-IT Trophy': ['HTC', '7 Trophy'],
  Spark: ['HTC', '7 Trophy'],
  'HTC 7 Trophy': ['HTC', '7 Trophy'],
  'HTC Trophy': ['HTC', '7 Trophy'],
  '(Radar|.*C110e)!': ['HTC', 'Radar'],
  Omega: ['HTC', 'Radar'],
  Mazaa: ['HTC', 'Mazaa'], // Never released
  Titan: ['HTC', 'Titan'],
  '.*X310e!': ['HTC', 'Titan'],
  PI39100: ['HTC', 'Titan'],
  Eternity: ['HTC', 'Titan'],
  Ultimate: ['HTC', 'Titan'],
  PI86100: ['HTC', 'Titan II'],
  '8X': ['HTC', '8X'],
  HTC6690LVW: ['HTC', '8X'],
  HTC6990LVW: ['HTC', '8X'],
  'HTC6990LVW 4G': ['HTC', '8X'],
  '6990LVW': ['HTC', '8X'],
  'RIO U': ['HTC', '8X'],
  'ACD U': ['HTC', '8X'],
  Accord: ['HTC', '8X'],
  'C620(d|e|t)!': ['HTC', '8X'],
  'C625(a|b)!': ['HTC', '8X'],
  '8X by HTC': ['HTC', '8X'],
  'Windows Phone 8X by HTC!': ['HTC', '8X'],
  WindowsPhone8Xby: ['HTC', '8X'],
  PM23300: ['HTC', '8X'],
  'htc 8x': ['HTC', '8X'],
  'A620(b|d|e|m|t)!': ['HTC', '8S'],
  '8S': ['HTC', '8S'],
  '8S by HTC': ['HTC', '8S'],
  'Windows Phone 8S by HTC!': ['HTC', '8S'],
  WindowsPhone8Sby: ['HTC', '8S'],
  HTCPO881: ['HTC', '8XT'],
  HTC6995LVW: ['HTC', 'One (M8) for Windows'],
  '0P6B140': ['HTC', 'One (M8) for Windows'],
  '0P6B180': ['HTC', 'One (M8) for Windows'],
  'HUAWEI Ascend W1!': ['Huawei', 'Ascend W1'],
  'HUAWEI ?W1-(C00|U00|U34)!': ['Huawei', 'Ascend W1'],
  'W1-(C00|U00|U34)!': ['Huawei', 'Ascend W1'],
  H883G: ['Huawei', 'Ascend W1'],
  'HUAWEI ?W2-(T00|T01|U00|U051)!': ['Huawei', 'Ascend W2'],
  'W2-(T00|T01|U00|U051)!': ['Huawei', 'Ascend W2'],
  '4Afrika': ['Huawei', '4Afrika'],
  Andi4L: ['iBall', 'Andi 4L Pulse'],
  M1010: ['Ingram Micro', 'M1010'],
  'Titanium Wind W4': ['Karbonn', 'Titanium Wind W4'],
  W4: ['Karbonn', 'Titanium Wind W4'],
  'Thunder 340W': ['KAZAM', 'Thunder 340W'],
  'Thunder 450W': ['KAZAM', 'Thunder 450W'],
  SOUL2: ['Kruger&Matz', 'Soul 2'],
  'iris Win 1': ['Lava', 'Iris Win 1'],
  Win1: ['Lava', 'Iris Win 1'],
  '- Lazer SMART WIND': ['Lazer', 'Smart Wind'], // By Archos
  GW910: ['LG', 'Optimus 7'],
  'LG-E740!': ['LG', 'Miracle'],
  'LG-E-?900!': ['LG', 'Optimus 7'],
  'LG-E906': ['LG', 'Jil Sander'],
  'LG-C900!': ['LG', 'Optimus 7Q'],
  VW820: ['LG', 'Lancet'],
  Quantum: ['LG', 'Quantum'],
  'Harley Davidson': ['NGM', 'Harley Davidson'],
  'Harley-Davidson': ['NGM', 'Harley Davidson'],
  NEO: ['NuAns', 'Neo'],
  'MegaFon SP-W1': ['Мегафон', 'SP-W1'],
  'MWP-47!': ['Miia', 'Windows Phone 4.7" iimotion'],
  W121: ['Micromax', 'Canvas Win W121'],
  'MICROMAX-W121': ['Micromax', 'Canvas Win W121'],
  M5w: ['Mint', 'M5 Windows'],
  TC70: ['Motorola', 'TC70'], // Motorola Solutions
  'MouseComputer MADOSMA Q501': ['MouseComputer', 'MADOSMA Q501'],

  '(Lumia ?|Nokia ?)?[0-9]{3,4}!!': {
    'Lumia 430!': ['Microsoft', 'Lumia 430'],
    'Lumia 435!': ['Microsoft', 'Lumia 435'],
    '(Lumia ?|Nokia ?)?505!': ['Nokia', 'Lumia 505'],
    '(Lumia ?|Nokia ?)?510!': ['Nokia', 'Lumia 510'],
    '(Lumia ?|Nokia ?)?520!': ['Nokia', 'Lumia 520'],
    '(Lumia ?|Nokia ?)?521!': ['Nokia', 'Lumia 521'],
    '(Lumia ?|Nokia ?)?525!': ['Nokia', 'Lumia 525'],
    '(Lumia ?|Nokia ?)?526!': ['Nokia', 'Lumia 526'],
    '(Lumia ?|Nokia ?)?530!': ['Nokia', 'Lumia 530'],
    'Lumia 532!': ['Microsoft', 'Lumia 532'],
    'Lumia 535!': ['Microsoft', 'Lumia 535'],
    'Lumia 540!': ['Microsoft', 'Lumia 540'],
    'Lumia 550!': ['Microsoft', 'Lumia 550'],
    '(Lumia ?|Nokia ?)?610!': ['Nokia', 'Lumia 610'],
    '(Lumia ?|Nokia ?)?620!': ['Nokia', 'Lumia 620'],
    '(Lumia ?|Nokia ?)?625!': ['Nokia', 'Lumia 625'],
    '(Lumia ?|Nokia ?)?630!': ['Nokia', 'Lumia 630'],
    '(Lumia ?|Nokia ?)?635!': ['Nokia', 'Lumia 635'],
    '(Lumia ?|Nokia ?)?636!': ['Nokia', 'Lumia 636'],
    'Lumia 638': ['Nokia', 'Lumia 638'],
    '(Lumia ?|Nokia ?)?640 XL!': ['Microsoft', 'Lumia 640 XL'],
    '(Lumia ?|Nokia ?)?640!': ['Microsoft', 'Lumia 640'],
    'Lumia 650!': ['Microsoft', 'Lumia 650'],
    '(Lumia ?|Nokia ?)?710!': ['Nokia', 'Lumia 710'],
    '(Lumia ?|Nokia ?)?719!': ['Nokia', 'Lumia 719'],
    '(Lumia ?|Nokia ?)?720!': ['Nokia', 'Lumia 720'],
    'Lumia 730!': ['Nokia', 'Lumia 730'],
    'Lumia 735!': ['Nokia', 'Lumia 735'],
    '(Lumia ?|Nokia ?)?800!': ['Nokia', 'Lumia 800'],
    '(Lumia ?|Nokia ?)?810!': ['Nokia', 'Lumia 810'],
    '(Lumia ?|Nokia ?)?820!': ['Nokia', 'Lumia 820'],
    '(Lumia ?|Nokia ?)?822!': ['Nokia', 'Lumia 822'],
    '(Lumia ?|Nokia ?)?830!': ['Nokia', 'Lumia 830'],
    'Lumia 850!': ['Microsoft', 'Lumia 850'],
    '(Lumia ?|Nokia ?)?900!': ['Nokia', 'Lumia 900'],
    '(Lumia ?|Nokia ?)?920!': ['Nokia', 'Lumia 920'],
    '(Lumia ?|Nokia ?)?925!': ['Nokia', 'Lumia 925'],
    '(Lumia ?|Nokia ?)?928!': ['Nokia', 'Lumia 928'],
    '(Lumia ?|Nokia ?)?929!': ['Nokia', 'Lumia 929 Icon'],
    '(Lumia ?|Nokia ?)?930!': ['Nokia', 'Lumia 930'],
    'Lumia 950 XL!': ['Microsoft', 'Lumia 950 XL'],
    'Lumia 950!': ['Microsoft', 'Lumia 950'],
    '(Lumia ?|Nokia ?)?909!': ['Nokia', 'Lumia 1020'], // 909 was the original name of the 1020
    '(Lumia ?|Nokia ?)?1020!': ['Nokia', 'Lumia 1020'],
    '(Lumia ?|Nokia ?)?1320!': ['Nokia', 'Lumia 1320'],
    '(Lumia ?|Nokia ?)?1520!': ['Nokia', 'Lumia 1520'],
    'Lumia 1530': ['Microsoft', 'Lumia 1530']
  },

  'RM-[0-9]{3,4}!!': {
    'RM-1099!': ['Microsoft', 'Lumia 430'],
    'RM-10(68|69|70|71)!': ['Microsoft', 'Lumia 435'],
    'RM-1114!': ['Microsoft', 'Lumia 435'],
    'RM-9(13|14|15)!': ['Nokia', 'Lumia 520'],
    'RM-917!': ['Nokia', 'Lumia 521'],
    'RM-998!': ['Nokia', 'Lumia 525'],
    'RM-997!': ['Nokia', 'Lumia 526'],
    'RM-10(17|18|19|20)!': ['Nokia', 'Lumia 530'],
    'RM-10(31|32|34)!': ['Microsoft', 'Lumia 532'],
    'RM-1115!': ['Microsoft', 'Lumia 532'],
    'RM-10(89|90|91|92)!': ['Microsoft', 'Lumia 535'],
    'RM-1141!': ['Microsoft', 'Lumia 540'],
    'RM-1127!': ['Microsoft', 'Lumia 550'],
    'RM-846!': ['Nokia', 'Lumia 620'],
    'RM-9(41|42|43)!': ['Nokia', 'Lumia 625'],
    'RM-9(76|77|78|79)!': ['Nokia', 'Lumia 630'],
    'RM-9(74|75)!': ['Nokia', 'Lumia 635'],
    'RM-1078!': ['Nokia', 'Lumia 635'],
    'RM-1027!': ['Nokia', 'Lumia 636'],
    'RM-1010!': ['Nokia', 'Lumia 638'],
    'RM-10(72|74|75|77)!': ['Microsoft', 'Lumia 640'],
    'RM-11(09|13)!': ['Microsoft', 'Lumia 640'],
    'RM-10(63|64|65|66|67|73|96)!': ['Microsoft', 'Lumia 640 XL'],
    'RM-11(52|54)!': ['Microsoft', 'Lumia 650'],
    'RM-8(85|87)!': ['Nokia', 'Lumia 720'],
    'RM-1040!': ['Nokia', 'Lumia 730'],
    'RM-1038!': ['Nokia', 'Lumia 735'],
    'RM-878!': ['Nokia', 'Lumia 810'],
    'RM-8(24|25|26)!': ['Nokia', 'Lumia 820'],
    'RM-845!': ['Nokia', 'Lumia 822'],
    'RM-9(83|84|85)!': ['Nokia', 'Lumia 830'],
    'RM-8(20|21|22)!': ['Nokia', 'Lumia 920'],
    'RM-867!': ['Nokia', 'Lumia 920'],
    'RM-8(92|93)!': ['Nokia', 'Lumia 925'],
    'RM-910!': ['Nokia', 'Lumia 925'],
    'RM-955!': ['Nokia', 'Lumia 925T'],
    'RM-860!': ['Nokia', 'Lumia 928'],
    'RM-927!': ['Nokia', 'Lumia 929 Icon'],
    'RM-10(45|87)!': ['Nokia', 'Lumia 930'],
    'RM-11(04|18)!': ['Microsoft', 'Lumia 950'],
    'RM-1085!': ['Microsoft', 'Lumia 950 XL'],
    'RM-1116!': ['Microsoft', 'Lumia 950 XL'],
    'RM-8(75|76|77)!': ['Nokia', 'Lumia 1020'],
    'RM-9(94|95|96)!': ['Nokia', 'Lumia 1320'],
    'RM-9(37|38|39|40)!': ['Nokia', 'Lumia 1520']
  },

  'id[0-9]{3,3}!!': {
    'id102!': ['Nokia', 'Lumia 822'],
    'id308!': ['Nokia', 'Lumia 1520'],
    'id313!': ['Nokia', 'Lumia 635'],
    'id316!': ['Nokia', 'Lumia 530']
  },

  SeaRay: ['Nokia', 'Lumia 800'],
  Semaphore: ['Nokia', '"Juggernaut Semaphore"'],
  'Surface Phone': ['Microsoft', 'Surface Phone'],
  'ONIX AWP4-215': ['ONYX', 'AWP4-215'],
  'QSMART STORM W408': ['Q-Mobile', 'Q-Smart Storm W408'],
  'QSMART DREAM W473': ['Q-Mobile', 'Q-Smart Dream W473'],
  'QSMART STORM W510': ['Q-Mobile', 'Q-Smart Storm W510'],
  'FZ-E1': ['Panasonic', 'Toughpad FZ-E1'],
  PSP8400DUO: ['Prestigio', 'MultiPhone 8400 Duo'],
  PSP8500DUO: ['Prestigio', 'MultiPhone 8500 Duo'],
  'PRESTIGIO-PSP8500DUO': ['Prestigio', 'MultiPhone 8500 Duo'],

  'GT-[IS][0-9]{4,4}!!': {
    'GT-I8350!': ['Samsung', 'Omnia W'],
    'GT-i8700': ['Samsung', 'Omnia 7'],
    'GT-I8710': ['Samsung', 'Ativ S'],
    'GT-I8750': ['Samsung', 'Ativ S'],
    'GT-S7530!': ['Samsung', 'Omnia M']
  },

  'SGH-[IT][0-9]{3,3}!!': {
    'SGH-I187': ['Samsung', 'Ativ S Neo'],
    'SGH-i667': ['Samsung', 'Focus 2'],
    'SGH-i677': ['Samsung', 'Focus Flash'],
    'SGH-i707': ['Samsung', 'Taylor'], // Developer phone - never released
    'SGH-i917!': ['Samsung', 'Focus'],
    'SGH-i937!': ['Samsung', 'Focus S'],
    'SGH-T899M': ['Samsung', 'Ativ S']
  },

  'SM-W750V': ['Samsung', 'Ativ SE'],
  'SPH-I800': ['Samsung', 'Ativ S Neo'],
  'SCH-R860U': ['Samsung', 'Ativ Odyssey'],
  'SCH-I930': ['Samsung', 'Ativ Odyssey'],
  'Ativ S': ['Samsung', 'Ativ S'],
  'ATIV S Neo': ['Samsung', 'Ativ S Neo'],
  CETUS: ['Samsung', 'Focus'],
  I917: ['Samsung', 'Focus'],
  'Focus i917!': ['Samsung', 'Focus'],
  'FOCUS S': ['Samsung', 'Focus S'],
  'OM(NI|IN)A ?7!': ['Samsung', 'Omnia 7'],
  'Omnia W': ['Samsung', 'Omnia W'],
  Haden: ['Samsung', 'Omnia W'],
  Taylor: ['Samsung', 'Taylor'],
  S606: ['TCL', 'Horizon S606'],
  TSUNAGI: ['Toshiba', 'Tsunagi'],
  TG01: ['Toshiba', 'TG01'],
  'WP 4.7': ['Trekstor', 'WP 4.7'],
  MI4: ['Xiaomi', 'Mi 4'],
  'Win-Q900S': ['XOLO', 'Win Q900s'],
  'Win Q1000': ['XOLO', 'Win Q1000'],
  WinQ1000: ['XOLO', 'Win Q1000'],
  'Billy 4': ['Yezz', 'Billy 4'],
  'Billy 4.7': ['Yezz', 'Billy 4.7'],
  '47': ['Yezz', 'Monaco 47'],
  USCCN859: ['ZTE', 'N859 Render'],
  'N880e!': ['ZTE', 'N880e'],
  V965W: ['ZTE', 'V965W Tania'],
  Tania: ['ZTE', 'V965W Tania'],

  'Microsoft!': [null, null],
  'Windows!': [null, null],
  'Virtual!': [null, null],
  'Android!': [null, null],
  'Linux!': [null, null],
  '<model>!': [null, null],
  'XXX!': [null, null],
  'InfoPath!': [null, null]
};

/* This file is automatically generated, do not edit manually! */
